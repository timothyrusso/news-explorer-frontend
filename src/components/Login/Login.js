import React, { useState } from 'react';
import PopupWithForm from '../PopupWithForm/PopupWithForm';
import Input from '../Input/Input';
import PopupServerError from '../PopupServerError/PopupServerError';
import { useDispatch } from 'react-redux';
import { setIsLoadingTextTrueAction } from '../../store/toggles/toggles.actions';
import { useCheckValidityInput } from '../../hooks/useCheckInputValidity';
import { useSelector } from 'react-redux';

const Login = ({
  isOpen,
  popupRedirectText,
  isLoadingText,
  formValidity,
  handleLoginSubmit,
  popupServerErrorMessage,
}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const errorMessage = useSelector((state) => state.errors.errorMessage);

  const { checkValidity } = useCheckValidityInput(errorMessage);

  const dispatch = useDispatch();

  const handleEmailChange = (evt) => {
    checkValidity(evt);
    setEmail(evt.target.value);
  };

  const handlePasswordChange = (evt) => {
    checkValidity(evt);
    setPassword(evt.target.value);
  };

  const handleSubmit = (evt) => {
    dispatch(setIsLoadingTextTrueAction());
    evt.preventDefault();
    handleLoginSubmit(password, email);
  };

  React.useEffect(() => {
    setEmail('');
    setPassword('');
  }, [isOpen]);

  return (
    <>
      <PopupWithForm
        name="signin"
        title="Sign in"
        isOpen={isOpen}
        popupRedirectText={popupRedirectText}
        buttonText={'Sign in'}
        loadingText={'Loading..'}
        isLoadingText={isLoadingText}
        onSubmit={handleSubmit}
        formValidity={formValidity}
      >
        <Input
          type={'email'}
          idName={'email-signin-input'}
          name={'email'}
          fieldName={'field_email'}
          placeholder={'Enter email'}
          value={email}
          onChange={handleEmailChange}
          errorMessage={errorMessage}
          labelText={'Email'}
        />
        <Input
          type={'password'}
          idName={'password-signin-input'}
          name={'password'}
          fieldName={'field_password'}
          placeholder={'Enter password'}
          minLength={'8'}
          maxLength={'20'}
          value={password}
          onChange={handlePasswordChange}
          errorMessage={errorMessage}
          labelText={'Password'}
        />
        <PopupServerError popupServerErrorMessage={popupServerErrorMessage} />
      </PopupWithForm>
    </>
  );
};

export default Login;
