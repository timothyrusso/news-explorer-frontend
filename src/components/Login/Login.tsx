import React from 'react';
import PopupWithForm from '../PopupWithForm/PopupWithForm';
import Input from '../Input/Input';
import PopupServerError from '../PopupServerError/PopupServerError';
import { useDispatch } from 'react-redux';
import { setIsLoadingTextTrueAction } from '../../store/toggles/toggles.actions';
import { useAuthenticationApi } from '../../hooks/useAuthenticationApi';
import { useAppSelector } from '../../hooks/useAppSelector';
import { useLoginForm } from '../../hooks/useLoginForm';
import { useInputValidity } from '../../hooks/useInputValidity';

const Login = () => {
  const errorMessage = useAppSelector((state) => state.errors.errorMessage);
  const isOpen = useAppSelector((state) => state.toggles.isSigninPopupOpen);
  const popupRedirectText = useAppSelector(
    (state) => state.toggles.popupRedirectText
  );
  const isLoadingText = useAppSelector((state) => state.toggles.isLoadingText);
  const popupServerErrorMessage = useAppSelector(
    (state) => state.errors.popupServerErrorMessage
  );

  const { email, setEmail, password, setPassword } = useLoginForm();
  const { handleLoginSubmit } = useAuthenticationApi(email, password);
  const { handleInputChange } = useInputValidity(errorMessage);

  const dispatch = useDispatch();

  const handleSubmit = (evt: React.FormEvent<HTMLFormElement>) => {
    dispatch(setIsLoadingTextTrueAction());
    evt.preventDefault();
    handleLoginSubmit();
  };

  React.useEffect(() => {
    setEmail('');
    setPassword('');
  }, [isOpen, setEmail, setPassword]);

  return (
    <>
      <PopupWithForm
        name="signin"
        title="Sign in"
        isOpen={isOpen}
        popupRedirectText={popupRedirectText}
        buttonText={'Sign in'}
        loadingText={'Loading..'}
        isLoadingText={isLoadingText}
        onSubmit={handleSubmit}
      >
        <Input
          type={'email'}
          idName={'email-signin-input'}
          name={'email'}
          fieldName={'field_email'}
          placeholder={'Enter email'}
          value={email}
          onChange={(evt) => handleInputChange(evt, setEmail)}
          errorMessage={errorMessage}
          labelText={'Email'}
        />
        <Input
          type={'password'}
          idName={'password-signin-input'}
          name={'password'}
          fieldName={'field_password'}
          placeholder={'Enter password'}
          minLength={8}
          maxLength={20}
          value={password}
          onChange={(evt) => handleInputChange(evt, setPassword)}
          errorMessage={errorMessage}
          labelText={'Password'}
        />
        <PopupServerError popupServerErrorMessage={popupServerErrorMessage} />
      </PopupWithForm>
    </>
  );
};

export default Login;
