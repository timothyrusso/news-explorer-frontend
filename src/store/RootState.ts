import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
