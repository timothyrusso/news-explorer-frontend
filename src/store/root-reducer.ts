import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
