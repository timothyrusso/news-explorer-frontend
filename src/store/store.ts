import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
