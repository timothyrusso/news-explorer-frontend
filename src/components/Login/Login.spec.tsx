import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store/store';
import Login from './Login';

describe('Login component', () => {
  it('Should render the Login component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
