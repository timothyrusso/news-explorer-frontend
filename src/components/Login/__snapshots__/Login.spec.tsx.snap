// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Should render the Login component 1`] = `
<DocumentFragment>
  <div
    class="popup  popup_type_signin"
  >
    <div
      class="popup__container"
    >
      <form
        action="#"
        class="popup__form"
        name="myFormmyFormsignin"
      >
        <h2
          class="popup__title"
        >
          Sign in
        </h2>
        <label
          class="popup__input-label"
          for="email-signin-input"
        >
          Email
        </label>
        <input
          class="popup__input popup__input_field_email"
          id="email-signin-input"
          name="email"
          placeholder="Enter email"
          required=""
          type="email"
          value=""
        />
        <span
          class="popup__input_type_error"
          id="email-signin-input-error"
        />
        <label
          class="popup__input-label"
          for="password-signin-input"
        >
          Password
        </label>
        <input
          class="popup__input popup__input_field_password"
          id="password-signin-input"
          maxlength="20"
          minlength="8"
          name="password"
          placeholder="Enter password"
          required=""
          type="password"
          value=""
        />
        <span
          class="popup__input_type_error"
          id="password-signin-input-error"
        />
        <p
          aria-live="polite"
          class="popup-server-error"
        />
        <button
          aria-label="Submit Sign in"
          class="submit-button popup__button  "
          type="submit"
        >
          Sign in
        </button>
        <p
          class="form-redirect"
        >
          or 
          <button
            aria-label="test"
            class="form-redirect_type_link"
            type="button"
          />
        </p>
      </form>
      <button
        aria-label="Close signin popup"
        class="close-button "
        type="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
